### PR TITLE
Reduce array abstraction on apple platforms dealing with literals

### PR DESCRIPTION
**Pull request by lattner**
Part of the ongoing quest to reduce swift array literal abstraction
penalties: make the SIL optimizer able to eliminate bridging overhead
when dealing with array literals.

Introduce a new classify_bridge_object SIL instruction to handle the
logic of extracting platform specific bits from a Builtin.BridgeObject
value that indicate whether it contains a ObjC tagged pointer object,
or a normal ObjC object. This allows the SIL optimizer to eliminate
these, which allows constant folding a ton of code. On the example
added to test/SILOptimizer/static_arrays.swift, this results in 4x
less SIL code, and also leads to a lot more commonality between linux
and apple platform codegen when passing an array literal.

This also introduces a couple of SIL combines for patterns that occur
in the array literal passing case.

This is the same as PR13659 but rebased on top of master.

### CODE REVIEW

**Comment by lattner** [comment]
@swift-ci please test

**Comment by lattner** [comment]
@swift-ci please test

**Comment by lattner** [comment]
I realize that the motivation for this may not be entirely obvious.  The point of this is to allow the SIL optimizer to reason about the magic bits that are mangled into the Builtin.BridgeObject values, instead of having them be extracted with and's of magic constants that are unknown to the optimizer (only IRGen does and should have access to this info).  Here is one example of a before and after this patch:

https://gist.githubusercontent.com/lattner/aa5cdefb9d9fe2b9f3508619e20b5225/raw/a6d7f82ded27760315219f9105d9e8f0119e0bec/classify-bridge-object.md

I recognize that it is theoretically possible that lazy bridging will go away.  However, until and if that happens, we should take this patch and continue providing more abstraction for the bridge representation.

**Comment by lattner** [comment]
I realize that the motivation for this may not be entirely obvious. The point of this is to allow the SIL optimizer to reason about the magic bits that are mangled into the Builtin.BridgeObject values, instead of having them be extracted with and's of magic constants that are unknown to the optimizer (only IRGen does and should have access to this info). Here is one example of a before and after this patch:

https://gist.github.com/lattner/aa5cdefb9d9fe2b9f3508619e20b5225

I recognize that it is theoretically possible that lazy bridging will go away. However, until and if that happens, we should take this patch and continue providing more abstraction for the bridge representation.

**Comment by lattner** [comment]
@swift-ci please test

**Comment by lattner** [comment]
@swift-ci please test

**Comment by lattner** [comment]
@swift-ci please test

**Review by eeckstein — COMMENTED** [review]
very nice! In general lgtm.

some comments: I think it makes  sense to clean up Builtin.swift a bit - to use the new builtin and remove some functions/variables which are not needed anymore now.

It would be nice to get rid of the bridge_object_to_word instruction at all. But afaik, it is still needed at least for getting spare bits out of the array buffer pointer (the needsElementTypeCheck flag). Either we extend classify_bridge_object returns to return this spare bit as well or we'll add another instruction. But we can do this in a follow-up work.

**Comment by jckarter** [comment]
Making this a high-level SIL operation makes a lot of sense. cc @milseman, who was experimenting with potential changes to BridgeObject to support the new String implementation. I think this is compatible with his plans.

**Comment by milseman** [comment]
This looks compatible with my use. 👍

File: docs/SIL.rst [hunk 1]
```diff
@@ -4617,6 +4617,19 @@ pointer_to_thin_function
...
 TODO
 
+classify_bridge_object
+``````````````````````
+::
+
+  sil-instruction ::= 'classify_bridge_object' sil-operand
+
+  %1 = classify_bridge_object %0 : $Builtin.BridgeObject
+  // %1 will be of type (Builtin.Int1, Builtin.Int1)
```
[the remark this answers is not in the document]
Comment (lattner) [reply docs/SIL.rst]: I took a look at that, but it requires a ton of boilerplate to define a multi-result instruction (and there are only a few of them that do).  If you feel strongly about this, I'd prefer to land this, then work on fixing the core representation of multi-result instructions in a follow-on patch, and adopt it later.

File: test/SILOptimizer/static_arrays.swift [hunk 2]
```diff
@@ -159,3 +159,33 @@ print(gg!)
...
+  print(ptr, len)  // Use the arguments somehow so they don't get removed.
+}
+
+// This should be a single basic block, and the array should end up being stack
+// allocated.
+//
+// CHECK-LABEL: sil @{{.*}}passArrayOfClasses
+// CHECK: bb0(%0 : $SwiftClass, %1 : $SwiftClass, %2 : $SwiftClass):
+// CHECK-NOT: bb1(
+// CHECK: alloc_ref {{.*}}[tail_elems $SwiftClass *
```
[the remark this answers is not in the document]
Comment (lattner) [reply test/SILOptimizer/static_arrays.swift]: Yes, exactly, and that is super annoying.  Not all targets end up having it be stack allocated in all configurations.  I think when the stdlib has checks enabled in particular it doesn't end up getting stack allocated.  This is all the class of things I'm interested in improving.

**Comment by lattner** [comment]
Thanks all, and yes, I consider this to be one step forward.  Eliminating bridge_object_to_word entirely is a dream :-)

File: docs/SIL.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjmccall) [reply docs/SIL.rst]: I'd be interested to know what you feel is broken about multi-result instructions.  I tried several different designs and concluded that there are just inherent trade-offs that have to be made here.  I chose to make working with single-result instructions easier at the cost of adding boilerplate to multi-result instructions.  If you have a concrete alternative design in mind, I'd love to hear it.

**Comment by jckarter** [comment]
We would like to keep around a way to get the raw payload of a BridgeObject with the trivial bit set. For the case of String in particular we'd use that ability for small string optimization. If we have to change `bridge_object_to_word` so that it's only defined behavior when the "tagged" bit is set, that'd probably be OK.

**Comment by gottesmm** [comment]
> On Jan 2, 2018, at 6:41 PM, John McCall <notifications@github.com> wrote:
> 
> @rjmccall commented on this pull request.
> 
> In docs/SIL.rst:
> 
> > @@ -4617,6 +4617,19 @@ pointer_to_thin_function
>  
>  TODO
>  
> +classify_bridge_object
> +``````````````````````
> +::
> +
> +  sil-instruction ::= 'classify_bridge_object' sil-operand
> +
> +  %1 = classify_bridge_object %0 : $Builtin.BridgeObject
> +  // %1 will be of type (Builtin.Int1, Builtin.Int1)
> 
> I'd be interested to know what you feel is broken about multi-result instructions. I tried several different designs and concluded that there are just inherent trade-offs that have to be made here. I chose to make working with single-result instructions easier at the cost of adding boilerplate to multi-result instructions. If you have a concrete alternative design in mind, I'd love to hear it.

Also, we added a helper mixin to make it really easy to define one. It is something like 10 lines of code.

**Pull request merged**